File: 206-reverse-linked-list/206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr) 
        {
            return nullptr;
        }
        
        if (head->next == nullptr)
        {
            return head;
        }
        
        ListNode* pNode = head;
        ListNode* pointer = head->next;
        while (pointer)
        {
            ListNode* tempPrevNext = pointer;
            ListNode* tempPointerNext = pointer->next;
            if (pNode == head)
            {
                pNode->next = nullptr;
            }
            pointer->next = pNode;
            pNode = tempPrevNext;
            pointer = tempPointerNext;
        }
        return pNode;
    }
};

File: 206-reverse-linked-list/206-reverse-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    if (head == null) {
        return null;
    }
        
    if (head.next == null){
        return head;
    }
        
    let pNode = head;
    let pointer = head.next;
    while (pointer){
        let tempPrevNext = pointer;
        let tempPointerNext = pointer.next;
        if (pNode == head){
            pNode.next = null;
        }
        pointer.next = pNode;
        pNode = tempPrevNext;
        pointer = tempPointerNext;
    }
    return pNode;
};